### PR TITLE
Add capture checking to some standard library classes

### PR DESCRIPTION
**Pull request by odersky**
Convert some standard library classes to capture checking. The converted classes are represented as tests.
Initially, the `Iterator` and `IterableOnce` classes are converted. UPDATE: By now, we have quite a few more, including List, ListBuffer, View, and all their super classes and traits.

Also change capture checker to facilitate the conversion. The main changes are about better interop with classes that are not (yet) capture checked.

Based on #18131

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala [hunk 1]
```diff
@@ -362,7 +362,15 @@ class CheckCaptures extends Recheck, SymTransformer:
         val argType0 = f(recheckStart(arg, pt))
         val argType = super.recheckFinish(argType0, arg, pt)
         super.recheckFinish(argType, tree, pt)
-      if meth == defn.Caps_unsafeBox then
+
+      if meth == defn.Caps_unsafeAssumePure then
+        val arg :: Nil = tree.args: @unchecked
+        val argType0 = recheck(arg, pt.capturing(CaptureSet.universal))
+        val argType = if argType0.captureSet.isAlwaysEmpty then argType0
+        else argType0.widen.stripCapturing
```
Comment (Linyxus) [inline compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala:402]: Indenting this line would make the conditional expression more readable

File: tests/pos-custom-args/captures/stdlib/collection/IterableOnce.scala [hunk 2]
```diff
@@ -84,7 +87,7 @@ trait IterableOnce[+A] extends Any {
 
 final class IterableOnceExtensionMethods[A](private val it: IterableOnce[A]) extends AnyVal {
```
Comment (Linyxus) [inline tests/pos-custom-args/captures/stdlib/collection/IterableOnce.scala]: Should `it` be impure?

File: tests/pos/stdlib/collection/IterableOnce.scala [hunk 3]
```diff
@@ -0,0 +1,1364 @@
...
+    s.asInstanceOf[S]
+  }
+
+  /** @return The number of elements in this $coll, if it can be cheaply computed,
+   *  -1 otherwise. Cheaply usually means: Not requiring a collection traversal.
+   */
+  def knownSize: Int = -1
+}
+
+final class IterableOnceExtensionMethods[A](private val it: IterableOnce[A]) extends AnyVal {
```
Comment (Linyxus) [inline tests/pos/stdlib/collection/IterableOnce.scala:88]: Should `it` be impure? otherwise the extension methods only apply on pure `IterableOnce`s.

Comment (odersky) [reply tests/pos/stdlib/collection/IterableOnce.scala:88]: I left it like that since all extension methods are deprecated anyway,

**Comment by odersky** [comment]
Observations of shortcomings so far:

 -In some instances we will get "cannot establish reference" errors when forming a capture set. For instance in the `stringOps` test in Test_2.scala.

 - Sometimes we'd like to have mutable variables capturing `cap`, but this is forbidden. We could do better by maintaining different universal capabilities at each nesting level, and only forbidding instantiating variables with universal capabilities at deeper level.

 - We have to assume that arrays are over sealed type variables

 - We have to ensure transitivity of sealed: A sealed type variable can only be instantiated with other sealed type variables or else with type variables upper bounded by deeply pure types. If there's level checking this has to be integrated.

**Comment by odersky** [comment]
@Linyxus This is about as far as I want to push it for now. Can you give it a rough review? I think you need to look only at the changes to the compiler itself; the test programs don't matter so much. It would be good to get it in fast since there are a lot of accumulated changes and other stuff we do with capture checking should build on these.

**Comment by Linyxus** [comment]
Thanks for the heads-up. I'll do the review right away.

**Review by Linyxus — APPROVED** [review]
LGTM!

**Comment by Linyxus** [comment]
So it seems that `test-windows-full` is failing for this PR (which is somehow skipped in the CI of this PR but run after merged into main). The error looks a bit weird: syntax error for files that look like [this](https://github.com/lampepfl/dotty/blob/main/tests/pos/stdlib/immutable_Iterable.scala). Not sure what this means.

/cc @odersky

**Comment by odersky** [comment]
These are symlinks. Maybe Windows does not know about them and interprets them literally?

**Comment by odersky** [comment]
I pushed a new PR that should fix the problem.

**Pull request merged**